### PR TITLE
[SPARK-11700] [SQL] Remove thread local SQLContext  in SparkPlan

### PR DESCRIPTION
**Pull request by davies**
In 1.6, we introduce a public API to have a SQLContext for current thread, SparkPlan should use that.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 1]
```diff
@@ -45,9 +45,10 @@ import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.execution.ui.{SQLListener, SQLTab}
 import org.apache.spark.sql.sources.BaseRelation
 import org.apache.spark.sql.types._
-import org.apache.spark.sql.{execution => sparkexecution}
 import org.apache.spark.sql.util.ExecutionListenerManager
+import org.apache.spark.sql.{execution => sparkexecution}
 import org.apache.spark.util.Utils
+import org.apache.spark.{SparkContext, SparkException}
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: super nit: import order

**Comment by zsxwing** [comment]
LGTM

**Comment by zsxwing** [comment]
There are other places need to be fixed.

**Comment by rxin** [comment]
Can you update the pull request title to say more?

"use weak reference to avoid memory leak" is a bit too generic. The original jira title is ok, or a better one is to mention SQLContext.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 1 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: How to define this order in Intelij?

**Comment by davies** [comment]
@zsxwing The one in SQLContext have a `clearActive` APIs, so I think don't need WeakRef there.

**Comment by zsxwing** [comment]
> @zsxwing The one in SQLContext have a clearActive APIs, so I think don't need WeakRef there.

Sorry. I meant the compilation failed.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 2]
```diff
@@ -401,7 +402,7 @@ class SQLContext private[sql](
    */
   @Experimental
   def createDataFrame[A <: Product : TypeTag](rdd: RDD[A]): DataFrame = {
-    SparkPlan.currentContext.set(self)
+    SparkPlan.currentContext.set(new WeakReference[SQLContext](self))
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: A `WeakReference` is quite weak; you'll lose the reference on a full GC. Is this not going to cause the code to fail suddenly? I don't know that references are any remedy for a leak.

**Comment by davies** [comment]
@zsxwing @srowen I changed this to use setActive()/getActiveContext(),

**Comment by zsxwing** [comment]
LGTM

**Pull request closed without merging**